### PR TITLE
Add a mesh refinement criterion based on the compaction length

### PR DESCRIPTION
**Pull request by jdannberg**
This is useful in models with melt transport, as the compaction length is a typical length scale for these models.

### CODE REVIEW

File: source/mesh_refinement/compaction_length.cc [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+              fe_values.reinit(cell);
+              in.reinit(fe_values, cell, this->introspection(), this->get_solution(), true);
+              this->get_material_model().evaluate(in, out);
+
+              MaterialModel::MeltOutputs<dim> *melt_out = out.template get_additional_output<MaterialModel::MeltOutputs<dim> >();
+              AssertThrow(melt_out != NULL,
+                          ExcMessage("Need MeltOutputs from the material model for computing the melt properties."));
+
+              // for each temperature dof, check if the compaction length exceeds the cell size
+              for (unsigned int i=0; i<this->get_fe().base_element(this->introspection().base_elements.temperature).dofs_per_cell; ++i)
```
Comment (tjhei) [inline source/mesh_refinement/compaction_length.cc]: you use the compositional field support points to initialize the quadrature, but you use the temperature support points here, is that on purpose? That looks incorrect.

File: source/mesh_refinement/compaction_length.cc [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    ASPECT_REGISTER_MESH_REFINEMENT_CRITERION(CompactionLength,
+                                              "compaction length",
+                                              "A mesh refinement criterion for models with melt transport "
+                                              "that computes refinement indicators based on the compaction "
+                                              "length, defined as "
+                                              "$\\delta = \\sqrt{\\frac{(\\xi + 4 \\eta/3) k}{\\eta_f}}$. "
+                                              "$\\xi$ is the bulk viscosity, $\\eta$ is the shear viscosity, "
+                                              "$k$ is the permeability and $\\eta_f$ is the melt viscosity. "
+                                              "If the cell width or height exceeds a multiple (which is "
+                                              "specified as an input parameter) of this compaction length , "
```
Comment (tjhei) [inline source/mesh_refinement/compaction_length.cc]: remove space before comma

File: tests/compaction_length_refinement.prm [hunk 3]
```diff
@@ -0,0 +1,140 @@
+# Listing of Parameters
```
Comment (tjhei) [inline tests/compaction_length_refinement.prm]: remove these two lines

**Comment by tjhei** [comment]
Neat! This is going to be useful.

File: source/mesh_refinement/compaction_length.cc (hunk 1 above)
Comment (jdannberg) [reply source/mesh_refinement/compaction_length.cc]: You're right, I copy/pasted it together from different places... I am using the composition everywhere now, which probably makes most sense.

**Comment by jdannberg** [comment]
Thanks for the review! Should all be fixed now.

File: include/aspect/mesh_refinement/compaction_length.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+
+    /**
+     * A class that implements a mesh refinement criterion based on the
+     * compaction length, a typical length scale in models with melt
+     * migration. Every cell that contains a value exceeding the cell
+     * width/height is marked for refinement.
```
Comment (bangerth) [inline include/aspect/mesh_refinement/compaction_length.h]: I don't think I understand this last sentence. Can you clarify?

File: source/mesh_refinement/compaction_length.cc [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Mesh refinement");
+      {
+        prm.enter_subsection("Compaction length");
+        {
+          prm.declare_entry("Mesh cells per compaction length", "1.0",
+                            Patterns::Double (0),
+                            "The ratio between compaction length and size of the mesh cells, "
+                            "or, in other words, how many cells the mesh should haver per "
```
Comment (bangerth) [inline source/mesh_refinement/compaction_length.cc]: haver -> have

File: source/mesh_refinement/compaction_length.cc [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    CompactionLength<dim>::
+    declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Mesh refinement");
+      {
+        prm.enter_subsection("Compaction length");
+        {
+          prm.declare_entry("Mesh cells per compaction length", "1.0",
+                            Patterns::Double (0),
+                            "The ratio between compaction length and size of the mesh cells, "
```
Comment (bangerth) [inline source/mesh_refinement/compaction_length.cc]: maybe say "The DESIRED ratio..."

File: source/mesh_refinement/compaction_length.cc [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                            "or, in other words, how many cells the mesh should haver per "
+                            "compaction length.");
```
Comment (bangerth) [inline source/mesh_refinement/compaction_length.cc]: I think you should explain in this description how this parameter is used, e.g., "This mesh refinement criterion will mark cells for refinement if ..."

File: source/mesh_refinement/density.cc [hunk 8]
```diff
@@ -59,11 +59,6 @@ namespace aspect
                                quadrature,
                                update_quadrature_points | update_values | update_gradients);
 
-      // the values of the compositional fields are stored as block vectors for each field
-      // we have to extract them in this structure
-      std::vector<std::vector<double> > prelim_composition_values (this->n_compositional_fields(),
-                                                                   std::vector<double> (quadrature.size()));
-
```
Comment (bangerth) [inline source/mesh_refinement/density.cc]: for future cases, make this into a separate PR. this is logically independent and would be easy to merge.

**Comment by bangerth** [comment]
Looks good. Please squash and merge once the tester is happy.

**Pull request merged**